### PR TITLE
perf(ddtrace/tracer): add WithTags to reduce per-tag closure allocations

### PR DESCRIPTION
**Pull request by darccio**
## Summary

`Tag(k, v)` allocates one closure per call, and contribs that set several tags per span (valkey, mongo) rebuild that whole option chain on every request even though most of the tags (component, span kind, db system, target host/port, etc.) never change for the lifetime of the client. Profiling on a high-throughput edge service attributed a meaningful share of the tracer's allocations to this pattern.

- Adds `WithTags(map[string]any) StartSpanOption` to `ddtrace/tracer` — merges a tag map into `StartSpanConfig.Tags` in one call instead of one `Tag()` closure per entry. It composes with the existing `WithStartSpanConfig`, which already does map-merging.
- Migrates the highest-volume contribs to build a base `StartSpanConfig` with their **constant** tags once at client/monitor construction, and merge the **per-request** tags via `WithTags`:
  - `contrib/valkey-io/valkey-go`
  - `contrib/go.mongodb.org/mongo-driver/mongo`
  - `contrib/go.mongodb.org/mongo-driver.v2/mongo`
- Left `gin-gonic/gin` alone — it already pre-builds its static `spanOpts` once per `Middleware()` call and copies it per-request via `options.Expand`, so it wasn't in the same class of problem.

### A correctness gotcha found along the way

`WithStartSpanConfig` merges `Tags` as:
```go
if c.Tags == nil {
    c.Tags = cfg.Tags       // aliases, does not copy
} else if cfg.Tags != nil {
    maps.Copy(c.Tags, cfg.Tags)
}
```
If `WithStartSpanConfig(base)` runs *before* any other tag-setting option in the chain, the per-call config's `Tags` gets aliased directly to the shared `base.Tags` map (not copied). A later `Tag()`/`WithTags()` call in that same chain would then mutate the shared, concurrently-reused base map — a real data race for a per-client cached base config.

The safe pattern is to always put `WithTags(dynamicMap)` (or any `Tag()` call) *before* `WithStartSpanConfig(base)`, so the per-call map already exists and the base is only ever read via `maps.Copy`, never aliased. This is documented on `WithTags`'s doc comment and exercised by `TestWithTagsBeforeWithStartSpanConfigDoesNotMutateBase` in `option_test.go`. All three migrated contribs follow this ordering.

## Benchmarks

A same-package synthetic benchmark showed ~0 difference — Go's inliner/escape-analysis erases the `Tag()` closures entirely when caller and callee share a package and args are literals, so it doesn't reproduce the real cross-package cost.

`BenchmarkStartSpan` in `contrib/valkey-io/valkey-go/valkey_test.go` reproduces the actual contrib-vs-tracer package boundary (real `tracer.Start`/`StartSpanFromContext`), comparing the old per-tag `Tag()` code (kept as `startSpanLegacy` for comparison) against the new `client.startSpan`:

```
BenchmarkStartSpan/legacy_tag_per_call-10          99 allocs/op   11288 ns/op
BenchmarkStartSpan/with_tags_and_static_base-10    92 allocs/op    9490 ns/op
```
(-7 allocs/op, ~16% faster, Apple M1 Max, `-benchtime=100000x`)

## Test plan

- [x] `make apidiff` — `WithTags` reported as the only, compatible change
- [x] `go test ./ddtrace/tracer/...` — green (one pre-existing goleak-only failure from idle agent HTTP connections, confirmed present on unmodified `main` too, see CONTRIBUTING.md's Goroutine Leaks section)
- [x] `go vet` + `gofmt -l` clean on all changed files
- [x] `go build ./...` for the whole workspace
- [x] Real Docker-backed integration tests for all 3 migrated contribs (`INTEGRATION=1 go test ./...` against the `valkey` and `mongodb` docker-compose services) — all pass

### CODE REVIEW

File: contrib/valkey-io/valkey-go/valkey_test.go [hunk 1]
```diff
@@ -438,3 +439,69 @@ func TestNewClient(t *testing.T) {
...
+		opts = append(opts, tracer.Tag(ext.TargetHost, c.host))
+	}
+	if c.port != "" {
+		opts = append(opts, tracer.Tag(ext.TargetPort, c.port))
+	}
+	if c.user != "" {
+		opts = append(opts, tracer.Tag(ext.DBUser, c.user))
+	}
+	return tracer.StartSpanFromContext(ctx, "valkey.command", opts...)
+}
```
Comment (rarguelloF) [inline contrib/valkey-io/valkey-go/valkey_test.go]: This can be removed and the benchmark results can be compared by commit instead

File: ddtrace/tracer/option.go [hunk 2]
```diff
@@ -1176,6 +1176,25 @@ func Tag(k string, v any) StartSpanOption {
...
+// set several tags per span on a hot path (e.g. per-request contrib tags).
+//
+// When used together with WithStartSpanConfig, WithTags must appear first in
+// the option list so its map becomes the owned, per-call Tags map; calling it
+// after WithStartSpanConfig would instead merge into (and mutate) the shared
+// base config's Tags map.
+func WithTags(tags map[string]any) StartSpanOption {
+	return func(cfg *StartSpanConfig) {
+		if cfg.Tags == nil {
+			cfg.Tags = tags
```
[the remark this answers is not in the document]
Comment (rarguelloF) [reply ddtrace/tracer/option.go]: please review, this should be addressed

File: ddtrace/tracer/option_test.go [hunk 3]
```diff
@@ -1954,6 +1954,72 @@ func TestWithStartSpanConfig(t *testing.T) {
 	assert.Equal(tm.UnixNano(), s.start)
 }
 
+func TestWithTags(t *testing.T) {
```
Comment (rarguelloF) [inline ddtrace/tracer/option_test.go:2139]: nit: these could be collapsed into a single test with 3 subtests for better organization

**Review by rarguelloF — COMMENTED** [review]
nice improvements! only some minor comments/nits

File: ddtrace/tracer/option_test.go (hunk 3 above)
Comment (darccio) [reply ddtrace/tracer/option_test.go:2139]: Done in 53b5db8

File: contrib/valkey-io/valkey-go/valkey_test.go (hunk 1 above)
Comment (darccio) [reply contrib/valkey-io/valkey-go/valkey_test.go]: Done in 53b5db8

File: ddtrace/tracer/option.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (darccio) [reply ddtrace/tracer/option.go]: Done in 53b5db8

**Review by rarguelloF — APPROVED** [review]

**Comment by darccio** [comment]
@codex review

File: contrib/IBM/sarama/dispatcher.go [hunk 4]
```diff
@@ -43,30 +69,29 @@ func (w *wrappedDispatcher) Run() {
...
 		}
 		if len(w.cfg.consumerCustomTags) > 0 {
 			for tag, tagValueFn := range w.cfg.consumerCustomTags {
-				opts = append(opts, tracer.Tag(tag, tagValueFn(msg)))
+				tags[tag] = tagValueFn(msg)
 			}
 		}
+		opts := []tracer.StartSpanOption{
+			tracer.WithTags(tags),
+			tracer.WithStartSpanConfig(w.spanCfg),
```
[the remark this answers is not in the document]
Comment (darccio) [reply contrib/IBM/sarama/dispatcher.go]: Good catch — fixed in 7c858517d for both the consumer (`dispatcher.go`) and producer (`producer.go`) paths: custom tags now go through a separate `WithTags` call placed *after* `WithStartSpanConfig(spanCfg)`, so they win on key collision again, matching pre-migration behavior. Added `TestWrappedDispatcherCustomTagPrecedence` and `TestStartProducerSpanCustomTagPrecedence`, both confirmed (via `git stash`) to fail against the pre-fix code and pass after.

While checking for the same pattern elsewhere in this PR, `contrib/confluentinc/confluent-kafka-go/kafkatrace`'s `WithCustomTag`/`tagFns` had the identical bug (not flagged here, but same root cause) — fixed in fddc47f54.

File: ddtrace/tracer/option.go [hunk 5]
```diff
@@ -1181,6 +1181,26 @@ func Tag(k string, v any) StartSpanOption {
...
+// When used together with WithStartSpanConfig, place WithTags (and any Tag
+// call) before WithStartSpanConfig in the option list: WithStartSpanConfig
+// aliases the base config's Tags map when no per-call tags have been set
+// yet, so a tag write after it would mutate the shared base config's map.
+func WithTags(tags map[string]any) StartSpanOption {
+	return func(cfg *StartSpanConfig) {
+		if cfg.Tags == nil {
+			cfg.Tags = make(map[string]any, len(tags))
+		}
+		maps.Copy(cfg.Tags, tags)
```
[the remark this answers is not in the document]
Comment (darccio) [reply ddtrace/tracer/option.go]: Addressed in e2c1cacb8 — added a new CONTRIBUTING.md section ("Favor a cached `StartSpanConfig` and `WithTags` over per-operation `Tag` closures") covering when to use `Tag` vs `WithTags` vs `NewStartSpanConfig`/`WithStartSpanConfig`, the ordering hazard between them (with a regression test proving it), and measured `allocs/op` from a benchmark parametrized on real contrib call-site shapes (n=1,2,4,6 dynamic tags).

**Comment by darccio** [comment]
@codex review

**Comment by darccio** [comment]
@codex review

File: contrib/confluentinc/confluent-kafka-go/kafkatrace/consumer.go [hunk 6]
```diff
@@ -51,32 +51,56 @@ func WrapConsumeEventsChannel[E any, TE Event](tr *Tracer, in chan E, consumer C
...
-			opts = append(opts, tracer.Tag(key, tagFn(msg)))
+			customTags[key] = tagFn(msg)
 		}
-	}
-	if !math.IsNaN(tr.analyticsRate) {
-		opts = append(opts, tracer.Tag(ext.EventSampleRate, tr.analyticsRate))
+		// Applied last so a custom tag wins over both the cached static
+		// base and the tags above on key collision, matching pre-migration
+		// behavior where custom-tag options were appended last.
+		opts = append(opts, tracer.WithTags(customTags))
```
[the remark this answers is not in the document]
Comment (darccio) [reply contrib/confluentinc/confluent-kafka-go/kafkatrace/consumer.go]: Good catch — fixed in 47a8485f9: pulled `ext.EventSampleRate` out of the cached `consumerSpanCfg` base and re-applied it as a separate `Tag()` call after the `customTags` `WithTags` call in `StartConsumeSpan`, so it's unconditionally last again and wins any collision with `WithCustomTag(ext.EventSampleRate, ...)` — matching pre-migration behavior. Added `TestStartConsumeSpanAnalyticsRatePrecedence`, confirmed (via `git stash`) to fail against the pre-fix code and pass after.

Checked IBM/sarama for the same pattern: it doesn't have this regression, since its pre-migration order already had `analyticsRate` appended before the custom-tag loop, so custom tags already won that collision both before and after this PR.

File: ddtrace/tracer/option.go [hunk 7]
```diff
@@ -1219,6 +1219,26 @@ func Tag(k string, v any) StartSpanOption {
...
```
[the remark this answers is not in the document]
Comment (darccio) [reply ddtrace/tracer/option.go]: Good catch — fixed in b2c463721: `WithTags` now clones its input map immediately when called (before returning the option), instead of only copying it when the option is applied inside `StartSpan`. Same allocation cost as before — still one map alloc + copy, just moved earlier — and the clone is never nil even for a nil/empty input, so it doesn't reintroduce the `WithStartSpanConfig` aliasing footgun for that edge case. Added a regression test that mutates the input map after calling `WithTags` but before starting the span, confirming the span observes the value captured at call time.

**Comment by darccio** [comment]
@autotest review

File: ddtrace/tracer/option.go [hunk 8]
```diff
@@ -1219,6 +1219,35 @@ func Tag(k string, v any) StartSpanOption {
...
+	// Snapshot immediately, never nil even for a nil/empty input, so a
+	// later WithStartSpanConfig(base) in the same option list always sees a
+	// non-nil, distinct Tags map and merges into it (copy) instead of
+	// aliasing base's map — see
+	// TestWithStartSpanConfigAliasesCachedBaseWhenCalledFirst.
+	snapshot := make(map[string]any, len(tags))
+	maps.Copy(snapshot, tags)
+	return func(cfg *StartSpanConfig) {
+		if cfg.Tags == nil {
+			cfg.Tags = snapshot
```
[the remark this answers is not in the document]
Comment (darccio) [reply ddtrace/tracer/option.go]: Confirmed, fixed in aee05c291. The nil-Tags branch now always allocates a fresh copy of the snapshot for `cfg.Tags` instead of aliasing the snapshot directly, so a cached `WithTags` option applied to multiple spans no longer shares a map that a later `Tag()` call can mutate in place.

Added a regression test (`TestWithTags/cached_option_does_not_leak_across_spans`) that reuses a single `WithTags` option across two `StartSpan` calls with an intervening `Tag()` — it fails against the old code and passes now.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -298,6 +298,51 @@ When possible, prioritize creating or using internal implementations for repetit
 [fmt.Sprintf](https://pkg.go.dev/fmt#Sprintf) can introduce unnecessary overhead when building a string. Favor [string builders](https://pkg.go.dev/strings#Builder), or simple string concatenation, `a + "b" + c` over `fmt.Sprintf` when possible, especially in hot paths.
 Sample PR: <https://github.com/DataDog/dd-trace-go/pull/3365>
 
+### Favor a cached `StartSpanConfig` and `WithTags` over per-operation `Tag` closures
```
Comment (kakkoyun) [inline CONTRIBUTING.md]: I don't think, we need this in contributing docs.

File: ddtrace/tracer/option.go [hunk 10]
```diff
@@ -1219,6 +1219,39 @@ func Tag(k string, v any) StartSpanOption {
...
+// be safely reused (or mutated) by the caller once WithTags returns. It is
+// equivalent to calling Tag for every entry of tags, but avoids allocating
+// one closure per tag, which matters when a caller needs to set several
+// tags per span on a hot path (e.g. per-request contrib tags).
+//
+// When used together with WithStartSpanConfig, place WithTags (and any Tag
+// call) before WithStartSpanConfig in the option list: WithStartSpanConfig
+// aliases the base config's Tags map when no per-call tags have been set
+// yet, so a tag write after it would mutate the shared base config's map.
+func WithTags(tags map[string]any) StartSpanOption {
```
Comment (kakkoyun) [inline ddtrace/tracer/option.go:1234]: Is this used anywhere? Maybe we should have those changes in this PR as well.

**Review by kakkoyun — CHANGES_REQUESTED** [review]

File: ddtrace/tracer/option.go (hunk 10 above)
Comment (darccio) [reply ddtrace/tracer/option.go:1234]: It's used in the different PRs stacked on top of this one:

https://github.com/DataDog/dd-trace-go/pull/5198
https://github.com/DataDog/dd-trace-go/pull/5199
https://github.com/DataDog/dd-trace-go/pull/5200
https://github.com/DataDog/dd-trace-go/pull/5201

File: CONTRIBUTING.md (hunk 9 above)
Comment (darccio) [reply CONTRIBUTING.md]: Done in f96c0b796

**Review by kakkoyun — APPROVED** [review]
LGTM
